File: web/src/lib/channels.ts
import { api, loadApi, type ApiResult } from '$lib/api';
import type { ChannelPresence } from '$lib/protocol';

export type ChannelKind = 'text' | 'voice';
export type AutoplayOrder = 'ordered' | 'shuffled' | 'smart';

export interface ChannelMember {
	id: string;
	displayName: string;
	avatarUrl?: string;
}

export interface ChannelAutoplay {
	enabled: boolean;
	order: AutoplayOrder;
	/** One of the crew's playlists; absent when none is chosen. */
	playlistId?: string;
}

/**
 * A crew's channel (ADR-0058): a text channel is a name, a gate and a
 * scrollback; a voice channel is a name, a gate, a deck and who is in it.
 * The list holds only the channels you may enter — a private one that does
 * not name you is not in it at all. The crew's owner and admins keep them
 * (#2434); a private one admits them and the members named into it.
 */
export interface CrewChannel {
	id: string;
	kind: ChannelKind;
	name: string;
	position: number;
	private: boolean;
	/** A voice channel's; a text channel makes no sound. */
	soundPack?: 'base' | 'silent';
	autoplay?: ChannelAutoplay;
	/** A private channel's named members; owner and admins enter by role. */
	members?: ChannelMember[];
	/** A voice channel's: who is in it right now (#2436). */
	presence?: ChannelPresence;
}

/** Absent fields keep their value; `playlistId: ''` chooses none. */
export interface ChannelPatch {
	name?: string;
	position?: number;
	private?: boolean;
	soundPack?: 'base' | 'silent';
	autoplay?: Partial<ChannelAutoplay>;
}

export function fetchCrewChannels(
	crewId: string,
	fetcher: typeof fetch = fetch,
): Promise<ApiResult<{ channels: CrewChannel[] }>> {
	return loadApi(fetcher, `/api/crews/${crewId}/channels`);
}

/** Where a text channel is read. */
export const textChannelPath = (crewId: string, channelId: string) =>
	`/crew/${crewId}/c/${channelId}`;

/** The create action's words, short: a rider's text channel is a chat (#2696). */
export const newLabel = (kind: ChannelKind) =>
	kind === 'text' ? 'New chat' : 'New voice';

/** The delete action's words: a rider's text channel is a chat channel (#2696). */
export const deleteLabel = (kind: ChannelKind) =>
	kind === 'text' ? 'Delete the chat' : 'Delete the channel';

/** Where a voice channel is (#2449). */
export const voiceChannelPath = (crewId: string, channelId: string) =>
	`/crew/${crewId}/v/${channelId}`;

export function createChannel(
	crewId: string,
	channel: { kind: ChannelKind; name: string; private?: boolean },
): Promise<ApiResult<CrewChannel>> {
	return api<CrewChannel>(`/api/crews/${crewId}/channels`, {
		method: 'POST',
		json: channel,
	});
}

export function updateChannel(
	id: string,
	patch: ChannelPatch,
): Promise<ApiResult<CrewChannel>> {
	return api<CrewChannel>(`/api/channels/${id}`, {
		method: 'PATCH',
		json: patch,
	});
}

/** Takes everything in it — chat, play log, recaps. Ask first (errors.md). */
/** What deleting a channel takes, for the confirm that asks first (errors.md).
 *  A private voice channel's plans go with it (#2610); an open one's stay on
 *  the crew's schedule. */
export function deleteChannelWarning(c: {
	kind: ChannelKind;
	private?: boolean;
}): string {
	if (c.kind === 'text')
		return 'Every message and image in it goes with it, for everyone in the crew. There is no undo.';
	const plans = c.private
		? 'Any session planned in it is cancelled, and whoever could ride it is told.'
		: 'Any session planned in it stays on the crew’s schedule, with no voice channel.';
	return `Its play log and the recaps of the sessions ridden in it go with it, for everyone in the crew. ${plans} There is no undo.`;
}

export function deleteChannel(id: string): Promise<ApiResult<void>> {
	return api<void>(`/api/channels/${id}`, { method: 'DELETE' });
}

/** Name a crew member into a private channel, or take them out of it. */
export function setNamedInChannel(
	id: string,
	userId: string,
	named: boolean,
): Promise<ApiResult<void>> {
	return api<void>(`/api/channels/${id}/members/${userId}`, {
		method: named ? 'PUT' : 'DELETE',
	});
}

/** One line a coach marked, as every surface that draws it reads it. */
export interface Announcement {
	/** The marked message, so the strip can point back at the line. */
	messageId: string;
	text: string;
	/** The message's author, not whoever marked it. */
	from: string;
	/** The author's id, for their status beside the name (ADR-0060). */
	fromId?: string;
	/** ISO — the message's own timestamp, not the marking's. */
	at: string;
}

/** A text channel's marked line, as the crew's Board leads with it. */
export interface CrewAnnouncement extends Announcement {
	channelId: string;
	channelName: string;
}

/**
 * The newest announcement across the crew's text channels the caller may
 * enter (ADR-0058) — undefined when none is up, the Board's normal state.
 */
export function fetchCrewAnnouncement(
	crewId: string,
	fetcher: typeof fetch = fetch,
): Promise<ApiResult<CrewAnnouncement | undefined>> {
	return loadApi(fetcher, `/api/crews/${crewId}/announcement`);
}

File: server/internal/channels/channels.go
// Package channels is a crew's text and voice channels (ADR-0058, #2434):
// list, create, rename, reorder, gate and delete them, and name members into
// a private one. Durable data only. What is live in a voice channel — who is
// in the call, the session running there — is the hub's (#2436, #2438).
package channels

import (
	"log/slog"
	"net/http"
	"time"

	"github.com/jackc/pgx/v5/pgtype"

	"github.com/natrontech/wattroom/server/internal/httpx"
	"github.com/natrontech/wattroom/server/internal/protocol"
	"github.com/natrontech/wattroom/server/internal/store"
	"github.com/natrontech/wattroom/server/internal/store/db"
)

// UserSource is what channels needs from auth — defined here, where it is
// consumed, and satisfied by *auth.Service.
type UserSource interface {
	RequireUser(w http.ResponseWriter, r *http.Request, signInMessage string) (db.User, bool)
	User(r *http.Request) (db.User, bool)
}

// Live is what channels asks of the hub, which keys by voice channel (#2436).
// Satisfied by *hub.Hub. Optional: without it nobody hears a change until
// they reload, and a voice row lists nobody in it.
type Live interface {
	// The lobby ping (#570): a channel changes because somebody else changed
	// it, and the ping is how every other client hears and re-fetches.
	PresenceChanged()
	// Who is in a voice channel right now.
	Presence(channel string) protocol.ChannelPresence
	// The session running in a voice channel, if one is (#2438).
	LiveSession(channel string) (protocol.LiveSession, bool)
	// Taking somebody out of a private channel severs them there too.
	Kick(channel, userID string)
	// A deleted voice channel's live state dies with it (#618).
	CloseRoom(channel string)
}

// VoiceEjector is LiveKit's half of a kick — satisfied by *av.Service, and
// absent without AV.
type VoiceEjector interface {
	Eject(channel, userID string)
}

// Canceller tells a cancelled plan's riders it is off (notify.Service).
type Canceller interface {
	SessionCancelled(crew, channel pgtype.UUID, workoutName string, startsAt time.Time, actor pgtype.UUID)
}

type Service struct {
	store    *store.Store
	users    UserSource
	log      *slog.Logger
	live     Live
	voice    VoiceEjector
	notifier Canceller
}

func New(st *store.Store, users UserSource, log *slog.Logger) *Service {
	return &Service{store: st, users: users, log: log}
}

// SetLive wires the hub in after construction — the hub needs this service
// first, as its door.
func (s *Service) SetLive(l Live) { s.live = l }

// SetVoiceEjector wires LiveKit ejection in when AV is configured.
func (s *Service) SetVoiceEjector(v VoiceEjector) { s.voice = v }

// SetNotifier wires the mail in when it is configured.
func (s *Service) SetNotifier(n Canceller) { s.notifier = n }

// evict severs somebody's live presence in one voice channel: the socket and
// the call.
func (s *Service) evict(channel, userID string) {
	if s.live != nil {
		s.live.Kick(channel, userID)
	}
	if s.voice != nil {
		s.voice.Eject(channel, userID)
	}
}

// changed pings every lobby socket; the ping carries no data.
// ponytail: one ping for every crew, not just this crew's members — the lobby
// has no per-crew routing yet (#2444), and a channel edit is a rare event.
func (s *Service) changed() {
	if s.live != nil {
		s.live.PresenceChanged()
	}
}

func (s *Service) Register(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/crews/live", s.handleCrewsLive)
	mux.HandleFunc("GET /api/crews/{id}/channels", s.handleList)
	mux.HandleFunc("GET /api/crews/{id}/live", s.handleLive)
	mux.HandleFunc("POST /api/crews/{id}/channels", s.handleCreate)
	mux.HandleFunc("PATCH /api/channels/{id}", s.handleUpdate)
	mux.HandleFunc("DELETE /api/channels/{id}", s.handleDelete)
	mux.HandleFunc("PUT /api/channels/{id}/members/{userID}", s.handleNameMember)
	mux.HandleFunc("DELETE /api/channels/{id}/members/{userID}", s.handleUnnameMember)
}

type memberJSON struct {
	ID          string  `json:"id"`
	DisplayName string  `json:"displayName"`
	AvatarURL   *string `json:"avatarUrl,omitempty"`
}

type autoplayJSON struct {
	Enabled bool   `json:"enabled"`
	Order   string `json:"order"`
	// A playlist of the crew's; absent when none is chosen.
	PlaylistID string `json:"playlistId,omitempty"`
}

type channelJSON struct {
	ID       string `json:"id"`
	Kind     string `json:"kind"`
	Name     string `json:"name"`
	Position int32  `json:"position"`
	Private  bool   `json:"private"`
	// A voice channel's; a text channel has no deck and makes no sound.
	SoundPack string        `json:"soundPack,omitempty"`
	Autoplay  *autoplayJSON `json:"autoplay,omitempty"`
	// A private channel's named members. The crew's owner and admins enter
	// by role and are not listed.
	Members []memberJSON `json:"members,omitempty"`
	// A voice channel's: who is in it right now (#2436).
	Presence *protocol.ChannelPresence `json:"presence,omitempty"`
}

func toJSON(c db.Channel, members []memberJSON) channelJSON {
	out := channelJSON{
		ID: store.UUIDString(c.ID), Kind: c.Kind, Name: c.Name,
		Position: c.Position, Private: c.Private,
	}
	if c.Kind == kindVoice {
		out.SoundPack = c.SoundPack
		out.Autoplay = &autoplayJSON{Enabled: c.AutoplayEnabled, Order: c.AutoplayOrder}
		if c.AutoplayPlaylistID.Valid {
			out.Autoplay.PlaylistID = store.UUIDString(c.AutoplayPlaylistID)
		}
	}
	if c.Private {
		out.Members = members
	}
	return out
}

// handleList: the crew's channels the caller may enter, text first. A private
// channel that does not name them is left out entirely rather than shown
// shut: a channel's existence is part of what its gate keeps (#2444 draws the
// same line for presence).
func (s *Service) handleList(w http.ResponseWriter, r *http.Request) {
	crewID, user, role, ok := s.crewFor(w, r)
	if !ok {
		return
	}
	rows, err := s.enterable(r.Context(), crewID, user.ID, role)
	if err != nil {
		httpx.Fail(w, s.log, "list channels failed", err, "The channels could not be loaded.", "crew", store.UUIDString(crewID))
		return
	}
	memberRows, err := s.store.Queries.ListChannelMembers(r.Context(), crewID)
	if err != nil {
		httpx.Fail(w, s.log, "list channel members failed", err, "The channels could not be loaded.", "crew", store.UUIDString(crewID))
		return
	}
	members := map[string][]memberJSON{}
	for _, m := range memberRows {
		id := store.UUIDString(m.ChannelID)
		members[id] = append(members[id], memberJSON{
			ID: store.UUIDString(m.ID), DisplayName: m.DisplayName, AvatarURL: m.AvatarUrl,
		})
	}
	out := []channelJSON{}
	for _, c := range rows {
		id := store.UUIDString(c.ID)
		entry := toJSON(c, members[id])
		if c.Kind == kindVoice && s.live != nil {
			presence := s.live.Presence(id)
			entry.Presence = &presence
		}
		out = append(out, entry)
	}
	httpx.WriteJSON(w, http.StatusOK, map[string]any{"channels": out})
}
